development/extensions/tutorial_testing.rst

```diff
@@ -0,0 +1,913 @@
+                ->with('acme_demo_goodbye')
+                ->willReturn(true);
+            $this->config->expects($this->once())
+                ->method('offsetGet')
+                ->with('acme_demo_goodbye')
+                ->willReturn($config);
+
+            $this->user->expects($this->once())
+                ->method('lang')
+                ->with($expected_language, $name)
+                ->willReturn($language_return);
+
+            $this->template->expects($this->once())
+                ->method('assign_var')
+                ->with('DEMO_MESSAGE', $language_return);
+
+            $this->helper->expects($this->once())
+                ->method('render')
+                ->with('demo_body.html',test_handle_bertie() $name, 200, false)
+                ->willReturn($this->getMock('Symfony\Component\HttpFoundation\Response'));
+
+            $response = $this->controller->handle($name);
+            $this->assertInstanceOf('\Symfony\Component\HttpFoundation\Response', $response);
+        }
+    }
+
+Using mocks
+-----------
+
+In the ``setUp()`` method we create our controller object. However, we do not
+use the actual classes of phpBB which are used by the controller, when opening
+the page. Instead
+`phpunit mocks <https://phpunit.de/manual/current/en/test-doubles.html>`_ are
+injected.
+
+These mocks allow to check how often a method is called, what the arguments are
+and let you specify the return value. This helps us to verify that **our**
+controller code behaves as expected. We do not want to get an error, when phpBB
+itself behaves unexpected.
+
+Testing a method
+----------------
+
+In our controller we added special handling when the provided name is
+``bertie``. So our first test is whether the
+``phpbb\controller\helper::message()`` method is called **once** with the
+``NO_AUTH_SPEAKING`` error message. We also want to make sure that the
+``phpbb\controller\helper::render()`` method is not called in this case.
+
+Then we call the controller's ``handle()`` method. If the methods are invoked
+correctly the test will pass.
+
+.. code-block:: php
+
+        public function test_handle_bertie()
+        {
+            $this->helper->expects($this->once())
+                ->method('message')
+                ->with('NO_AUTH_SPEAKING', array('bertie'), 'NO_AUTH_OPERATION', 403)
+                ->willReturn($this->getMock('Symfony\Component\HttpFoundation\Response'));
+
+            $this->helper->expects($this->never())
+                ->method('render');
+
+            $response = $this->controller->handle('bertie');
+            $this->assertInstanceOf('\Symfony\Component\HttpFoundation\Response', $response);
+        }
+
+.. note::
+
+    Make sure that the name of your testing method starts with ``test``.
+    Otherwise it is not being executed as a test by phpunit.
+
+Data providers
+--------------
+
+For the normal case we use a data provider. A data provider is a method that
```

Comment:
In most cases we will want to provide some test data to our unit test methods. A data provider is a method that returns an array of arrays containing input variables and expected output variables. The inner array contains the arguments for the ``test`` method.